### PR TITLE
SI: Forego device change delay for GC type change

### PR DESCRIPTION
**Pull request by malleoz**
NOTE: Clarification is needed (mostly for my sake) before this is considered for merging.

This PR aims to solve the following issue:
1. Set Gamecube port 1 as Standard Controller
2. Play Gamecube .dtm file
3. Load savestate created with Gamecube port 1 as WiiU Adapter
4. Notice 60 frames where inputs are not accepted/read from .dtm
5. After 60 frames, notice inputs being read from 60 frames earlier in the dtm than the current frame

This occurs because of the 1 second pause between controller switches within `SerialInterfaceManager::ChangeDeviceDeterministic`. **I'd like more clarification on which scenarios this applies to and whether the check I have in place is safe.**

My goal here is to eliminate the 1 second pause for scenarios in which the previous/next controllers are processed identically by the game. This issue makes TAS collaboration slightly more difficult, requiring every group member to maintain consistent controller emulation settings.

### CODE REVIEW

File: Source/Core/Core/HW/SI/SI.cpp [hunk 1]
```diff
@@ -509,9 +509,14 @@ void SerialInterfaceManager::ChangeDeviceDeterministic(SIDevices device, int cha
   if (m_channel[channel].has_recent_device_change)
     return;
 
-  if (GetDeviceType(channel) != SIDEVICE_NONE)
+  SIDevices previous_device = GetDeviceType(channel);
```
Comment (iwubcode) [inline Source/Core/Core/HW/SI/SI.cpp:512]: These two variables can be `const`

**Comment by Pokechu22** [comment]
My understanding is that the point of this delay is to let the game notice that the device changed (as otherwise there is no signal to it). [A similar delay exists for EXI devices](https://github.com/dolphin-emu/dolphin/blob/c3cc1de54e89f7d0dced58b247a95dd052b7c26e/Source/Core/Core/HW/EXI/EXI.cpp#L209-L220) (e.g. changing from memory cards to microphones).

Changing between an emulated GC controller and the Wii U adapter, though, shouldn't be something that games care about. The delay definitely isn't needed there. But I think that using `SIDevice_IsGCController` might cause issues - changing from a controller to the bongos is something that games need to detect and thus the device should temporarily be removed. (That's worth testing, though.)

I guess a related thing is that I'm pretty sure movies don't record device changes, so I don't know if it makes sense to allow changing devices at all while recording/playing back a movie. I thought something prevented it/automatically changed the settings in that case, and if that's true, then the temporary disconnection doesn't make sense at all.

**Comment by malleoz** [comment]
> My understanding is that the point of this delay is to let the game notice that the device changed (as otherwise there is no signal to it). [A similar delay exists for EXI devices](https://github.com/dolphin-emu/dolphin/blob/c3cc1de54e89f7d0dced58b247a95dd052b7c26e/Source/Core/Core/HW/EXI/EXI.cpp#L209-L220) (e.g. changing from memory cards to microphones).
> 
> Changing between an emulated GC controller and the Wii U adapter, though, shouldn't be something that games care about. The delay definitely isn't needed there. But I think that using `SIDevice_IsGCController` might cause issues - changing from a controller to the bongos is something that games need to detect and thus the device should temporarily be removed. (That's worth testing, though.)
> 
> I guess a related thing is that I'm pretty sure movies don't record device changes, so I don't know if it makes sense to allow changing devices at all while recording/playing back a movie. I thought something prevented it/automatically changed the settings in that case, and if that's true, then the temporary disconnection doesn't make sense at all.

Thanks for the reply. I didn't have a great understanding of which scenarios the delay was necessary, but your bongos example helps. In that case, I will only forego the delay if switching to/from Standard Controller and WiiU adapter.

Correct, movies do not record device changes. We should disallow controller changes during recording. I would however suggest that this shouldn't be enforced while playing back a movie. It seems inconvenient to have to adjust your controller settings accordingly in order to watch a TAS playback. I'd instead suggest that if you go from read-only playback to recording again, the controller should be enforced. I'll probably look into this as my next Dolphin contribution.

**Comment by malleoz** [comment]
Updated to only forego the delay when going to/from standard from/to WiiU adapter. This should be ready for merge then.

**Comment by jordan-woyak** [comment]
What about the differences in calibration data?

**Comment by malleoz** [comment]
Can you rephrase your question? I'm not following for some reason. My understanding is that controllers aren't re-calibrated during this switch, so they would retain whatever calibration was set before.

**Comment by jordan-woyak** [comment]
The "Wii-U Adapter" path uses the first pad state for calibration data. The "Standard Controller" path uses perfectly centered calibration data. Switching without informing the console is probably going to result in undesirable calibration data, but I don't know if there is really a good way to resolve that.
